### PR TITLE
GroupMetaFactoryArgs documentation

### PR DESCRIPTION
**Pull request by MHBauer**
**What this PR does / why we need it**:
Documentation for people writing new API-Groups.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: documentation

**Special notes for your reviewer**:
@deads2k @pmorie my thoughts from writing the service-catalog apiserver.

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: staging/src/k8s.io/apimachinery/pkg/apimachinery/announced/group_factory.go [hunk 1]
```diff
@@ -42,10 +42,16 @@ type GroupVersionFactoryArgs struct {
 
 // GroupMetaFactoryArgs contains the group-level args of a GroupMetaFactory.
 type GroupMetaFactoryArgs struct {
+	// GroupName is the name of the API-Group
```
Comment (MHBauer) [inline staging/src/k8s.io/apimachinery/pkg/apimachinery/announced/group_factory.go]: I wanted to write 'fully-qualified' or some such, but I don't see any consistency with existing apis.

```
λ rg 'const GroupName'
abac/register.go
26:const GroupName = "abac.authorization.kubernetes.io"

abac/v1beta1/register.go
25:const GroupName = "abac.authorization.kubernetes.io"

abac/v0/register.go
25:const GroupName = "abac.authorization.kubernetes.io"

apps/register.go
30:const GroupName = "apps"

apps/v1beta1/#register.go#
28:const GroupName = "apps"

apps/v1beta1/register.go
26:const GroupName = "apps"

authentication/#register.go#
27:const GroupName = "authentication.k8s.io"

authentication/register.go
25:const GroupName = "authentication.k8s.io"

authentication/v1beta1/register.go
26:const GroupName = "authentication.k8s.io"

authorization/register.go
25:const GroupName = "authorization.k8s.io"

authorization/v1beta1/register.go
26:const GroupName = "authorization.k8s.io"

autoscaling/register.go
25:const GroupName = "autoscaling"

autoscaling/v1/register.go
26:const GroupName = "autoscaling"

batch/register.go
25:const GroupName = "batch"

batch/v1/register.go
26:const GroupName = "batch"

batch/v2alpha1/register.go
26:const GroupName = "batch"

certificates/register.go
30:const GroupName = "certificates.k8s.io"

certificates/v1beta1/register.go
26:const GroupName = "certificates.k8s.io"

componentconfig/register.go
30:const GroupName = "componentconfig"

componentconfig/v1alpha1/register.go
25:const GroupName = "componentconfig"

extensions/register.go
25:const GroupName = "extensions"

extensions/v1beta1/register.go
26:const GroupName = "extensions"

imagepolicy/register.go
25:const GroupName = "imagepolicy.k8s.io"

imagepolicy/v1alpha1/register.go
26:const GroupName = "imagepolicy.k8s.io"

policy/register.go
25:const GroupName = "policy"

policy/v1alpha1/register.go
26:const GroupName = "policy"

policy/v1beta1/register.go
26:const GroupName = "policy"

rbac/register.go
25:const GroupName = "rbac.authorization.k8s.io"

rbac/v1alpha1/register.go
25:const GroupName = "rbac.authorization.k8s.io"

rbac/v1beta1/register.go
25:const GroupName = "rbac.authorization.k8s.io"

storage/register.go
25:const GroupName = "storage.k8s.io"

storage/v1beta1/register.go
26:const GroupName = "storage.k8s.io
```

File: staging/src/k8s.io/apimachinery/pkg/apimachinery/announced/group_factory.go [hunk 2]
```diff
@@ -42,10 +42,16 @@ type GroupVersionFactoryArgs struct {
...
+	// example: 'servicecatalog.k8s.io'
 	GroupName              string
 	VersionPreferenceOrder []string
-	ImportPrefix           string
-
+	// ImportPrefix is the base go package of the API-Group
+	//
+	// example: 'k8s.io/kubernetes/pkg/apis/autoscaling'
+	ImportPrefix string
+	// RootScopedKinds are resources that are not namespaced.
```
Comment (MHBauer) [inline staging/src/k8s.io/apimachinery/pkg/apimachinery/announced/group_factory.go]: This is a guess. Please confirm.

File: staging/src/k8s.io/apimachinery/pkg/apimachinery/announced/group_factory.go [hunk 3]
```diff
@@ -42,10 +42,16 @@ type GroupVersionFactoryArgs struct {
...
 // GroupMetaFactoryArgs contains the group-level args of a GroupMetaFactory.
 type GroupMetaFactoryArgs struct {
+	// GroupName is the name of the API-Group
+	//
+	// example: 'servicecatalog.k8s.io'
 	GroupName              string
 	VersionPreferenceOrder []string
-	ImportPrefix           string
-
+	// ImportPrefix is the base go package of the API-Group
```
Comment (MHBauer) [inline staging/src/k8s.io/apimachinery/pkg/apimachinery/announced/group_factory.go]: It always looks like a go package, but I do not know why.

**Comment by caesarxuchao** [comment]
cc @nikhiljindal

**Comment by nikhiljindal** [comment]
@k8s-bot ok to test

/lgtm

File: staging/src/k8s.io/apimachinery/pkg/apimachinery/announced/group_factory.go (hunk 2 above)
Comment (nikhiljindal) [inline staging/src/k8s.io/apimachinery/pkg/apimachinery/announced/group_factory.go]: Thats correct

**Comment by nikhiljindal** [comment]
cc @deads2k @sttts

File: staging/src/k8s.io/apimachinery/pkg/apimachinery/announced/group_factory.go (hunk 1 above)
Comment (deads2k) [inline staging/src/k8s.io/apimachinery/pkg/apimachinery/announced/group_factory.go]: History.  Mistakes were made early on, but we eventually settled on qualified.  In this repo: `*.k8s.io.`

File: staging/src/k8s.io/apimachinery/pkg/apimachinery/announced/group_factory.go (hunk 3 above)
Comment (deads2k) [inline staging/src/k8s.io/apimachinery/pkg/apimachinery/announced/group_factory.go]: > It always looks like a go package, but I do not know why.

There is/was some kind of reflective go-type lookup to build a hard coded RESTMapper.

File: staging/src/k8s.io/apimachinery/pkg/apimachinery/announced/group_factory.go (hunk 2 above)
Comment (MHBauer) [inline staging/src/k8s.io/apimachinery/pkg/apimachinery/announced/group_factory.go]: @nikhiljindal thank you.

File: staging/src/k8s.io/apimachinery/pkg/apimachinery/announced/group_factory.go (hunk 1 above)
Comment (MHBauer) [inline staging/src/k8s.io/apimachinery/pkg/apimachinery/announced/group_factory.go]: @deads2k Should I adjust the text, or do you think the example is clear enough?

File: staging/src/k8s.io/apimachinery/pkg/apimachinery/announced/group_factory.go (hunk 3 above)
Comment (MHBauer) [inline staging/src/k8s.io/apimachinery/pkg/apimachinery/announced/group_factory.go]: @deads2k thanks, I'll be on the lookout for that.

**Pull request merged**